### PR TITLE
Implement default DFS traversal visitor for LogQL expressions

### PR DESCRIPTION
**Pull request by chaudum**
**What this PR does / why we need it**:

This can be used as a default visitor for use cases where you only want
to visit specific expression without the needing to implement the
traversal logic of each and every expression, see implementation in test case.

**Checklist**
- [x] Reviewed the [`CONTRIBUTING.md`](https://github.com/grafana/loki/blob/main/CONTRIBUTING.md) guide (**required**)
- [ ] Documentation added
- [x] Tests updated
- [ ] `CHANGELOG.md` updated
  - [ ] If the change is worth mentioning in the release notes, add `add-to-release-notes` label
- [ ] Changes that require user attention or interaction to upgrade are documented in `docs/sources/setup/upgrade/_index.md`
- [ ] For Helm chart changes bump the Helm chart version in `production/helm/loki/Chart.yaml` and update `production/helm/loki/CHANGELOG.md` and `production/helm/loki/README.md`. [Example PR](https://github.com/grafana/loki/commit/d10549e3ece02120974929894ee333d07755d213)
- [ ] If the change is deprecating or removing a configuration option, update the `deprecated-config.yaml` and `deleted-config.yaml` files respectively in the `tools/deprecated-config-checker` directory. [Example PR](https://github.com/grafana/loki/pull/10840/commits/0d4416a4b03739583349934b96f272fb4f685d15)

### CODE REVIEW

File: pkg/logql/syntax/visit.go [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+	VisitLineFilter(*LineFilterExpr)
+	VisitLineFmt(*LineFmtExpr)
+	VisitLogfmtExpressionParser(*LogfmtExpressionParser)
+	VisitLogfmtParser(*LogfmtParserExpr)
+}
+
+var _ RootVisitor = &DepthFirstTraversal{}
+
+type DepthFirstTraversal struct {
+	VisitBinOpFn                  func(v *DepthFirstTraversal, e *BinOpExpr)
```
Comment (jeschkies) [inline pkg/logql/syntax/visit.go:48]: Would it be better to accept `RootVisitor` instead?

File: pkg/logql/syntax/visit.go [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+// VisitLogRange implements RootVisitor.
+func (v *DepthFirstTraversal) VisitLogRange(e *LogRange) {
+	if e == nil {
+		return
+	}
+	if v.VisitLogRangeFn != nil {
+		v.VisitLogRangeFn(v, e)
+	} else {
+		e.Left.Accept(v)
```
Comment (jeschkies) [inline pkg/logql/syntax/visit.go:204]: Only left? Is the another `LogRange` child?

**Review by jeschkies — APPROVED** [review]
Nice work!

File: pkg/logql/syntax/visit.go (hunk 2 above)
Comment (chaudum) [reply pkg/logql/syntax/visit.go:204]: Yes, there is only a Left, no other children.

**Pull request merged**